internal/webhook/inferenceservice/mutating.go

```diff
@@ -0,0 +1,212 @@
+			return validationResp
+		}
+
+		// only proceed with injection if the annotation is valid and shouldInject is true
+		if !shouldInject {
+			return admission.Allowed(fmt.Sprintf("Connection validation passed in namespace %s for %s, no injection needed", req.Namespace, req.Kind.Kind))
+		}
+
+		injectionPerformed, obj, err := w.performConnectionInjection(ctx, req, secret, connectionType)
+		if err != nil {
+			log.Error(err, "Failed to perform connection injection")
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		// finally, write updated object back to k8s
+		if injectionPerformed {
+			marshaledObj, err := json.Marshal(obj)
+			if err != nil {
+				return admission.Errored(http.StatusInternalServerError, err)
+			}
+			return admission.PatchResponseFromRaw(req.Object.Raw, marshaledObj)
+		}
+
+	default:
+		resp = admission.Allowed(fmt.Sprintf("Operation %s on %s allowed in namespace %s", req.Operation, req.Kind.Kind, req.Namespace))
+	}
+	return resp
+}
+
+func (w *ConnectionWebhook) performConnectionInjection(
+	ctx context.Context,
+	req admission.Request,
+	secret *corev1.Secret,
+	connectionType string,
+) (bool, *unstructured.Unstructured, error) {
+	log := logf.FromContext(ctx)
+
+	obj := &unstructured.Unstructured{}
+	if err := w.Decoder.Decode(req, obj); err != nil {
+		return false, nil, fmt.Errorf("failed to decode InferenceService object: %w", err)
+	}
+	log.Info("Decoded InferenceService object", "connectionType", connectionType, "operation", req.Operation)
+
+	// injection based on connection type
+	switch connectionType {
+	case "oci-v1":
+		if err := w.injectOCIImagePullSecrets(obj, secret.Name); err != nil {
+			return false, nil, fmt.Errorf("failed to inject OCI imagePullSecrets: %w", err)
+		}
+		log.Info("Successfully injected OCI imagePullSecrets", "secretName", secret.Name)
+		return true, obj, nil
+
+	case "uri-v1":
+		if err := w.injectURIStorageUri(obj, secret); err != nil {
+			return false, nil, fmt.Errorf("failed to inject URI storageUri: %w", err)
+		}
+		log.Info("Successfully injected URI storageUri from secret", "secretName", secret.Name)
+		return true, obj, nil
+
+	case "s3":
+		if err := w.injectS3StorageKey(obj, secret.Name); err != nil {
+			return false, nil, fmt.Errorf("failed to inject S3 storage.key: %w", err)
+		}
+		log.Info("Successfully injected S3 storage key", "secretName", secret.Name)
+		return true, obj, nil
+
+	default: // this should not enter, but keep it just for sanity check if allowedTypes is updated.
+		log.Info("Unknown connection type, skipping injection", "connectionType", connectionType)
+		return false, nil, nil
+	}
+}
+
+// injectOCIImagePullSecrets injects imagePullSecrets into spec.predictor.imagePullSecrets for OCI connections.
+func (w *ConnectionWebhook) injectOCIImagePullSecrets(obj *unstructured.Unstructured, secretName string) error {
+	imagePullSecrets, found, err := unstructured.NestedSlice(obj.Object, IsvcConfigs.ImagePullSecretPath...)
```

Comment:
maybe you can create a GetOrCreateNested slice here, so you can simplify the code below